x/contracts/wasmlanche/wasmlanche-test/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+
+            BUILD_SET.with_borrow_mut(|build_set| {
+                if build_set.contains(crate_name) {
+                    return;
+                }
+
+                let status = Command::new("cargo")
+                    .arg("rustc")
+                    .arg("--crate-type")
+                    .arg("cdylib")
+                    .arg("--package")
+                    .arg(crate_name)
+                    .arg("--target")
+                    .arg(WASM_TARGET)
+                    .arg("--profile")
+                    .arg(PROFILE)
+                    .arg("--target-dir")
+                    .arg(&target_dir)
+                    .status()
+                    .expect("cargo build failed");
+
+                if !status.success() {
+                    panic!("cargo build failed");
+                }
+
+                build_set.insert(crate_name.to_string());
+            });
+
+            target_dir
+                .join(WASM_TARGET)
+                .join(PROFILE)
+                .join(crate_name.replace('-', "_"))
+                .with_extension("wasm")
+        };
+
+        let mut config = Config::new();
+
+        config.cranelift_opt_level(OptLevel::Speed);
+        config.consume_fuel(true);
+        config.wasm_threads(false);
+        config.wasm_multi_memory(false);
+        config.wasm_memory64(false);
+        config.epoch_interruption(true); // I think this is redundant when there's fuel
+        config.cranelift_nan_canonicalization(true);
+        // config.static_memory_maximum_size(0); // 0 is dynamic-only
+
+        let engine = Engine::new(&config).expect("engine failure");
+
+        let module = WASM_CACHE.with_borrow_mut(|cache| {
+            let bytes = cache.entry(wasm_path).or_insert_with_key(|wasm_path| {
+                std::fs::read(wasm_path)
+                    .expect("failed to read wasm")
+                    .into_boxed_slice()
+            });
+
+            Module::from_binary(&engine, bytes).expect("failed to compile wasm")
+        });
+
+        let mut linker = Linker::new(&engine);
+
+        linker
+            .func_wrap(
+                "contract",
+                "set_call_result",
+                |mut caller: Caller<'_, StoreData>, ptr: u32, len: u32| {
+                    let Extern::Memory(memory) = caller
+                        .get_export("memory")
+                        .expect("memory should be exported")
+                    else {
+                        panic!("export `memory` should be of type `Memory`");
+                    };
+
+                    let (ptr, len) = (ptr as usize, len as usize);
+
+                    let result = memory
+                        .data(&caller)
+                        .get(ptr..ptr + len)
+                        .expect("data should exist")
+                        .to_vec();
+
+                    caller.data_mut().set_result(result);
+                },
+            )
```

Comment:
we could wrap `linker` into our own struct and then separate these `.func_wraps` into a separate file. Defining our imported modules & functions here makes the `Builder::new()` function a bit messy & hard to extend.

Reply:
I've got a fix for this, I'll add it as a separate commit

Reply:
1b4e109ccacfadfbe3250bd3222dd24d8a7087d4